File: 백준/Silver/2346. 풍선 터뜨리기/풍선 터뜨리기.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayDeque;
import java.util.Arrays;
import java.util.stream.Collectors;
import java.util.stream.IntStream;

public class Main {

    public static void main(String[] args) throws IOException {
        try (BufferedReader reader = new BufferedReader(new InputStreamReader(System.in))) {
            int n = Integer.parseInt(reader.readLine());
            int[] balloonValues = Arrays.stream(reader.readLine().split(" "))
                    .mapToInt(Integer::parseInt)
                    .toArray();

            System.out.print(solution(n, balloonValues));
        }
    }

    private static StringBuilder solution(int n, int[] values) {
        StringBuilder result = new StringBuilder();

        ArrayDeque<Balloon> deque = IntStream.range(0, n)
                .mapToObj(i -> new Balloon(i + 1, values[i]))
                .collect(Collectors.toCollection(ArrayDeque::new));

        while (!deque.isEmpty()) {
            Balloon balloon = deque.pollFirst();
            result.append(balloon.getIndex()).append(" ");

            if (balloon.turnRight()) {
                for (int i = 0; i < balloon.getTurnCount() - 1; i++) {
                    if (!deque.isEmpty()) {
                        deque.offerLast(deque.pollFirst());
                    }
                }
            } else {
                for (int i = 0; i < balloon.getTurnCount(); i++) {
                    if (!deque.isEmpty()) {
                        deque.offerFirst(deque.pollLast());
                    }
                }
            }
        }
        return result;
    }

    public static class Balloon {

        private final int index;
        private final int paperValue;


        public Balloon(int index, int paperValue) {
            this.index = index;
            this.paperValue = paperValue;
        }

        public boolean turnRight() {
            return 0 < paperValue;
        }

        public int getTurnCount() {
            return Math.abs(paperValue);
        }

        public int getIndex() {
            return index;
        }
    }
}

File: 백준/Silver/2346. 풍선 터뜨리기/풍선 터뜨리기.py
import sys
from collections import deque

input = sys.stdin.readline

# ----- 풍선 객체 -----
class Balloon:
    def __init__(self, index, paper_value):
        self.index = index
        self.paper_value = paper_value

    def turn_right(self):
        return self.paper_value > 0

    def get_turn_count(self):
        return abs(self.paper_value)

def solution(n, values):
    result = []
    balloon_deque = deque([Balloon(i + 1, values[i]) for i in range(n)])

    while balloon_deque:
        balloon = balloon_deque.popleft()
        result.append(balloon.index)

        if balloon.turn_right():
            for _ in range(balloon.get_turn_count() - 1):
                if balloon_deque:
                    balloon_deque.append(balloon_deque.popleft())
        else:
            for _ in range(balloon.get_turn_count()):
                if balloon_deque:
                    balloon_deque.appendleft(balloon_deque.pop())

    return result

# ----- 입력 & 출력 -----
n = int(input())
balloon_values = list(map(int, input().split()))
result = solution(n, balloon_values)

print(' '.join(map(str, result)))